LongWarOfTheChosen/Src/NewPromotionScreenByDefault_Integrated/Classes/NPSBDP_UIArmory_PromotionHero.uc

```diff
@@ -309,127 +309,157 @@ function bool HasBrigadierRank()
-	// So only do this if we have visible ability icons
-	while(`ISCONTROLLERACTIVE && !AllAbilityIconsHidden(Column) && !Column.AbilityIcons[Column.m_iPanelIndex].bIsVisible)
+	// XCom : If using a controller, select the first available/visible ability in the column.
+	// NPSBDP : It is possible for ranks to have no abilities if we offset them in a way that hides all ability icons;
+	// therefore, only do this if we have visible ability icons.
+	while (`ISCONTROLLERACTIVE && !AllAbilityIconsHidden(Column) && !Column.AbilityIcons[Column.m_iPanelIndex].bIsVisible)
 	{
 		Column.m_iPanelIndex +=1;
-		if( Column.m_iPanelIndex >= Column.AbilityIcons.Length )
+		if (Column.m_iPanelIndex >= Column.AbilityIcons.Length)
 		{
 			Column.m_iPanelIndex = 0;
 		}
 	}
-	// bsg-nlong (1.25.17): end
 
-	return bHasColumnAbility;
+	// KDM : The return value, HasColumnAbility, helps determine whether this column should have the 
+	// 'New Rank' graphic shown over top of its rank name and symbol. Unfortunately, the code above only
+	// checks whether a 'visible' ability has been chosen in this given column; it ignores abilities
+	// which are invisible due to scrolling. 
+	//
+	// Here is a simple example. Let's say we scroll down and are currently viewing the bottom 2 soldier
+	// ability rows, the XCom ability row, and the pistol ability row. Furthermore, let's assume that we 
+	// have already purchased the ability, within this column, located in the top soldier ability row. 
+	// Although HasColumnAbility should be true, due to having an ability in this column, it will, in fact, 
+	// be false since the invisible ability row is never checked.
+	//
+	// What we really want to do is, for this particular column, determine if one of its 'soldier abilities' 
+	// have been chosen/bought while, at the same time, ignoring 'XCom abilities' and 'pistol abilities'.
+	HasColumnAbility = false;
+	if (!ColumnAboveUnitRank)
+	{
+		// KDM : GetAbilitiesPerRank's return value is based upon the following criteria :
+		// 1.] If ClassAbilitiesPerRank has been defined for the unit's class, within XComPromotionUIMod.ini, then the
+		// associated AbilitiesPerRank value is returned. AbilitiesPerRank represents the number of 'soldier abilities'
+		// per rank for a particular soldier class.
+		// 2.] If ClassAbilitiesPerRank has NOT been defined for the unit's class then the function determines
+		// the number of abilities for each rank with respect to that class. The largest 'abilities for a given
+		// rank' value is then returned.
+		NumberOfSoldierAbilities = GetAbilitiesPerRank(Unit);
+		if (NumberOfSoldierAbilities > AbilityTree.Length)
+		{
+			NumberOfSoldierAbilities = AbilityTree.Length;
+		}
+
+		for (i = 0; i < NumberOfSoldierAbilities; i++)
+		{
+			AbilityTemplate = AbilityTemplateManager.FindAbilityTemplate(AbilityTree[i].AbilityName);
+			if (AbilityTemplate != none && Unit.HasSoldierAbility(AbilityTemplate.DataName))
```

Comment:
Why is the template needed? Would `Unit.HasSoldierAbility(AbilityTree[i].AbilityName)` not work?